concordium-cis2/src/lib.rs

```diff
@@ -1120,6 +1132,163 @@ pub struct OnReceivingCis2Params<T: IsTokenId, A: IsTokenAmount> {
+}
+
+impl StandardIdentifierOwned {
+    /// Validate and construct a standard identifier.
+    pub fn new(id: String) -> Result<Self, InvalidStandardIdentifierError> {
+        if id.len() > 255 || !id.is_ascii() {
+            Err(InvalidStandardIdentifierError)
+        } else {
+            Ok(Self {
+                id,
+            })
+        }
+    }
+
+    /// Construct a standard identifier without validation.
+    pub fn new_unchecked(id: String) -> Self {
+        Self {
+            id,
+        }
+    }
+
+    /// Convert to standard identifier.
+    pub fn as_standard_identifier(&self) -> StandardIdentifier {
+        StandardIdentifier::new_unchecked(&self.id)
+    }
+}
+
+/// The parameter type for the contract function `supports`.
+#[derive(Debug, Serialize)]
+pub struct SupportsQueryParams {
+    /// The list of support queries.
+    #[concordium(size_length = 2)]
+    pub queries: Vec<StandardIdentifierOwned>,
+}
+
+// This is implemented manually to flatten and simplify the schema.
+impl schema::SchemaType for SupportsQueryParams {
+    fn get_type() -> schema::Type {
+        schema::Type::List(schema::SizeLength::U16, Box::new(StandardIdentifierOwned::get_type()))
+    }
+}
+
+/// The query result type for whether a smart contract supports a standard.
+// Note: For the serialization to be derived according to the CIS0
+// specification, the order of the variants and their fields cannot be changed.
+#[derive(Debug, Serialize)]
+pub enum SupportResult {
+    /// The standard is not supported.
+    NoSupport,
+    /// The standard is supported by the current contract address.
+    Support,
+    /// The standard is supported by using another contract address.
+    SupportBy(#[concordium(size_length = 1)] Vec<ContractAddress>),
+}
+
+// This is implemented manually because another manual implementation depends on
+// this. Currently it is not practical to rely on a derived schema type in a
+// manual implemented schema type, because the derived version only generates
+// the implementation with a specific feature enabled.
+impl schema::SchemaType for SupportResult {
+    fn get_type() -> schema::Type {
+        schema::Type::Enum(vec![
+            ("NoSupport".to_string(), schema::Fields::None),
+            ("Support".to_string(), schema::Fields::None),
+            (
+                "SupportBy".to_string(),
+                schema::Fields::Unnamed(vec![schema::Type::List(
+                    schema::SizeLength::U8,
+                    Box::new(ContractAddress::get_type()),
+                )]),
+            ),
+        ])
+    }
+}
+
+/// The response which is sent back when calling the contract function
+/// `supports`. It consists of a list of results corresponding to the list of
+/// queries.
+#[derive(Debug, Serialize)]
+pub struct SupportsQueryResponse {
+    /// List of support results corresponding to the list of queries.
+    #[concordium(size_length = 2)]
+    pub results: Vec<SupportResult>,
+}
```

Comment:
All of these newtype wrappers should have `#[repr(transparent)]`.

We should also consider adding a `transparent` attribute to `derive(Schema)` to avoid these manual implementations as well.

We want to encourage type safety, and so we need to make newtype wrappers as easy as possible.